### PR TITLE
feat(datepicker): wire up selected value propagation

### PR DESCRIPTION
**Pull request by mmalerba**
as of this PR we have a bare minimum working datepicker \o/

### CODE REVIEW

File: src/lib/datepicker/datepicker-input.ts [hunk 1]
```diff
@@ -1,23 +1,78 @@
...
+  }
+  set value(value: any) {
+    this._value = this._locale.parseDate(value);
+    let stringValue = this._value == null ? '' : this._locale.formatDate(this._value);
+    this._renderer.setElementProperty(this._elementRef.nativeElement, 'value', stringValue);
+  }
+  private _value: SimpleDate;
+
+  // Implemented as part of ControlValueAccessor
+  writeValue(value: SimpleDate) {
```
Comment (kara) [inline src/lib/datepicker/datepicker-input.ts]: Return types

File: src/lib/datepicker/datepicker-input.ts [hunk 2]
```diff
@@ -1,23 +1,78 @@
...
+    return this._elementRef;
+  }
+
+  @Input()
+  get value() {
+    return this._value;
+  }
+  set value(value: any) {
+    this._value = this._locale.parseDate(value);
+    let stringValue = this._value == null ? '' : this._locale.formatDate(this._value);
```
Comment (kara) [inline src/lib/datepicker/datepicker-input.ts]: Nit: const?

File: src/lib/datepicker/index.ts [hunk 3]
```diff
@@ -20,8 +20,27 @@ export * from './year-view';
...
-  declarations: [MdCalendar, MdCalendarTable, MdDatepicker, MdDatepickerInput, MdMonthView, MdYearView],
+  imports: [
+    CommonModule,
+    DatetimeModule,
+    MdDialogModule,
+    OverlayModule
+  ],
+  exports: [
+    MdCalendar,
+    MdCalendarTable,
```
Comment (kara) [inline src/lib/datepicker/index.ts]: Do you intend for the CalendarTable, MonthView, etc. to be used externally? Seems like you can remove them from the exports list.

File: src/lib/datepicker/datepicker-input.ts (hunk 1 above)
Comment (mmalerba) [inline src/lib/datepicker/datepicker-input.ts]: I guess... I don't see much value in the `: void`

File: src/lib/datepicker/datepicker-input.ts [hunk 4]
```diff
@@ -1,23 +1,78 @@
...
+
+  constructor(private _elementRef: ElementRef, private _renderer: Renderer,
+              private _locale: CalendarLocale) {}
+
+  getPopupConnectionElementRef(): ElementRef {
+    return this._elementRef;
+  }
+
+  @Input()
+  get value() {
```
Comment (tinayuangao) [inline src/lib/datepicker/datepicker-input.ts]: nit: return type?

**Comment by mmalerba** [comment]
addressed comments, PTAL

File: src/lib/datepicker/datepicker-input.ts [hunk 5]
```diff
@@ -1,23 +1,87 @@
...
+  _onChange = (value: any) => {};
+
+  _onTouched = () => {};
+
+  constructor(private _elementRef: ElementRef, private _renderer: Renderer,
+              private _locale: CalendarLocale) {}
+
+  ngAfterContentInit() {
+    if (this._datepicker) {
+      this._datepicker.selectedChanged.subscribe((selected: SimpleDate) => {
```
Comment (kara) [inline src/lib/datepicker/datepicker-input.ts]: Do you unsubscribe from this anywhere?

File: src/lib/datepicker/datepicker.spec.ts [hunk 6]
```diff
@@ -122,3 +248,37 @@ class MultiInputDatepicker {
 class NoInputDatepicker {
   @ViewChild('d') datepicker: MdDatepicker;
 }
+
+
+@Component({
+  template: `
+    <input [mdDatepicker]="d" [value]="'1/1/2020'">
```
Comment (kara) [inline src/lib/datepicker/datepicker.spec.ts]: Nit: might be easier to scan if it's just `value="1/1/2020"`

File: src/lib/datepicker/datepicker.spec.ts [hunk 7]
```diff
@@ -95,14 +127,110 @@ describe('MdDatepicker', () => {
...
+      expect(testComponent.selected).toBeNull();
+      expect(testComponent.datepickerInput.value).toBeNull();
+
+      let selected = new SimpleDate(2017, 0, 1);
+      testComponent.datepicker.selected = selected;
+      detectModelChanges(fixture);
+
+      expect(testComponent.selected).toEqual(selected);
+      expect(testComponent.datepickerInput.value).toEqual(selected);
+    }));
```
Comment (kara) [inline src/lib/datepicker/datepicker.spec.ts]: We should add tests that ensure that touched, disabled, and dirty still work, given that we're re-implementing them in the value accessor

**Comment by mmalerba** [comment]
comments addressed

File: src/lib/datepicker/datepicker.spec.ts [hunk 8]
```diff
@@ -95,14 +129,159 @@ describe('MdDatepicker', () => {
...
+
+    it('should mark input dirty after date selected', fakeAsync(() => {
+      let inputEl = fixture.debugElement.query(By.css('input')).nativeElement;
+
+      expect(inputEl.classList).toContain('ng-pristine');
+
+      testComponent.datepicker.selected = new SimpleDate(2017, 0, 1);
+      detectModelChanges(fixture);
+
+      expect(inputEl.classList).toContain('ng-dirty');
```
Comment (kara) [inline src/lib/datepicker/datepicker.spec.ts]: It should actually not mark the control dirty based on programmatic changes. Dirtiness is a reflection of UI changes only (what the user does). We're probably calling the change callback somewhere we shouldn't be.

Comment (mmalerba) [inline src/lib/datepicker/datepicker.spec.ts]: This property isn't really intended to be set programmatically, it's set when the user clicks a calendar cell. Made it internal so people don't try to call it, though...

**Pull request merged**